apiserver/params/cloud.go

```diff
@@ -114,3 +114,67 @@ type CloudSpecResult struct {
 type CloudSpecResults struct {
 	Results []CloudSpecResult `json:"results,omitempty"`
 }
+
+// CloudCredentialRequestParam defines a credential in terms of its cloud and name.
+// It is used to request detailed content for the credential stored on the controller.
+type CloudCredentialRequestParam struct {
+	CloudName      string `json:"cloud-name"`
+	CredentialName string `json:"credential-name"`
+}
+
+func (p CloudCredentialRequestParam) IsEmpty() bool {
+	return p.CloudName == "" && p.CredentialName == ""
+}
+
+// CredentialContentRequestParam defines an input required to make a valid call
+// to get credentials content stored on the controller.
+type CredentialContentRequestParam struct {
+	Credentials    []CloudCredentialRequestParam `json:"credentials,omitempty"`
+	IncludeSecrets bool                          `json:"include-secrets"`
+}
+
+// CloudCredential contains a cloud credential content.
+type CredentialContent struct {
+	// Name is the short name of the credential.
+	Name string `json:"name"`
+
+	// Cloud is the cloud name to which this credential belongs.
+	Cloud string `json:"cloud"`
+
+	// AuthType is the authentication type.
+	AuthType string `json:"auth-type"`
+
+	// Attributes contains non-secret credential values.
+	Attributes map[string]string `json:"attrs,omitempty"`
+
+	// Secrets contains secret credential values.
+	Secrets map[string]string `json:"secrets,omitempty"`
+}
+
+// ModelAccess contains information about user model access.
+type ModelAccess struct {
+	Model  string `json:"model,omitempty"`
+	Access string `json:"access,omitempty"`
```

Comment:
s/string/UserAccessPermission

Reply:
So, this is the question that I've asked when doing state layer call... I could have returned state.Model and state.UserAccessPermission... Since I have not and you were ok with that, it's harder to construct anything but string at this layer... Should I revisit state?

Reply:
As discussed offline, this change has bigger repercussions and will be done as a follow-up and re-work of another area.